boat_simulator/nodes/data_collection/data_collection_node.py

```diff
@@ -0,0 +1,222 @@
+            msg_type_name = self.topics_to_record[i]
+
+            if msg_type_name not in self.__msg_types_dict:
+                self.get_logger().error(
+                    f"msg type {msg_type_name} does not exist. Please adjust the topics array in \
+                        globals.yaml"
+                )
+
+            self.__record_topics_sub.append(
+                create_subscription_with_callback(topic_name, self.__msg_types_dict[msg_type_name])
+            )
+
+            if self.use_json:
+                self.__data_to_write[topic_name] = None
+
+            if self.use_bag:
+                topic_info = rosbag2_py._storage.TopicMetadata(
+                    name=topic_name, type=msg_type_name, serialization_format="cdr"
+                )
+                self.__writer.create_topic(topic_info)
+
+    def __init_timer_callbacks(self):
+        """Initializes timer callbacks of this node that are executed periodically."""
+        self.get_logger().debug("Initializing timer callbacks...")
+        self.create_timer(timer_period_sec=self.json_write_period, callback=self.__write_to_json)
+
+    def __init_shutdown_callbacks(self):
+        """Initializes shutdown callbacks of this node that are executed on shutdown."""
+        self.get_logger().debug("Initializing shutdown callbacks...")
+        self.context.on_shutdown(self.__shutdown_callback)
+
+    # SUBSCRIPTION CALLBACKS
+    def __general_sub_callback(self, msg: Type, topic_name: str):
+        if self.use_json:
+            msg_as_ord_dict = rosidl_runtime_py.message_to_ordereddict(msg)
+            self.__data_to_write[topic_name] = msg_as_ord_dict
+
+        if self.use_bag:
+            self.__writer.write(
+                topic_name, serialize_message(msg), self.get_clock().now().nanoseconds
+            )
+
+    # TIMER CALLBACKS
+    def __write_to_json(self):
+        # Only concern with this is if topic subscribed to is not launched, then we will never
+        # write to json for all the other ones will never get written
+        if all(self.__data_to_write.values()):
+            time_in_seconds = (self.get_clock().now().nanoseconds - self.__start_time) / 1e9
+            self.__data_to_write["time"] = time_in_seconds
+            if self.__json_index_counter > 0:
+                self.__json_file.write(",\n")
+            item_to_write = {self.__json_index_counter: self.__data_to_write}
+            json.dump(item_to_write, self.__json_file, indent=4)
+            self.__json_index_counter += 1
+
+    # SHUTDOWN CALLBACKS
+    def __shutdown_callback(self):
+        """Shutdown callback to close bag and json."""
+        self.get_logger().debug("Closing the storage files...")
+
+        if self.use_json:
+            self.__json_file.write("\n]")  # Close the JSON array
+            self.__json_file.close()
+
+        if self.use_bag:
+            self.__writer.close()  # Needs to be called after json close to prevent early exit
```

Comment:
There seem to be many functions that have this pattern of checking if JSON or ROS bag is used by using two consecutive if statements. Is there a way to abstract this information using OOP (i.e. make a class that represents a writer object and then define the open, write, close, etc. operations for each writer?)

This could be an issue that someone takes on to improve the data collection node, but we could at least take note of it and create an issue at a later date.